### PR TITLE
[TS] LPS-169808 Add clock skew logic to LDAP import

### PR DESCRIPTION
**Pull request by markgulacsy**
Hi appsec team,

I was working on a new configuration option to handle the issue caused by the server clock time difference [LPS-169808](https://issues.liferay.com/browse/LPS-169808)
Could you please review this solution?

Thank you!

### CODE REVIEW

**Comment by markgulacsy** [comment]
ci:test:ldap

**Review by stian-sigvartsen — CHANGES_REQUESTED** [review]
Hi @markgulacsy ! Thanks for your PR. I have added some comments. I would like to bring most attention to the one about the infinite loop potential.

**Comment by stian-sigvartsen** [comment]
Hi @markgulacsy . To avoid any confusion about which team is reviewing this right now, I will close the PR and await a followup from yourself.

**Comment by markgulacsy** [comment]
Hi @stian-sigvartsen,
Thank you for your review and comments. I am checking your concerns and I will address them.

File: modules/apps/portal-security/portal-security-ldap-api/src/main/java/com/liferay/portal/security/ldap/exportimport/configuration/LDAPImportConfiguration.java [hunk 1]
```diff
@@ -35,6 +35,12 @@ public interface LDAPImportConfiguration extends CompanyScopedConfiguration {
 	@Override
 	public long companyId();
 
+	@Meta.AD(
+		deflt = "3000", description = "clock-skew-description",
```
[the remark this answers is not in the document]
Comment (markgulacsy) [reply modules/apps/portal-security/portal-security-ldap-api/src/main/java/com/liferay/portal/security/ldap/exportimport/configuration/LDAPImportConfiguration.java:39]: I do not understand why is this a breaking change. I think adding this property with a default value only influences the scenarios when the customer has issues with a future time coming from the LDAP server. Can you explain to me what am I missing here?

File: modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/exportimport/LDAPUserImporterImpl.java [hunk 2]
```diff
@@ -1740,6 +1740,24 @@ private User _updateUser(
...
+			Date expireDate = new Date(
+				System.currentTimeMillis() +
+					ldapImportConfiguration.clockSkew());
+
+			if (modifiedDate.compareTo(expireDate) > 0) {
+				throw new RuntimeException(
+					StringBundler.concat(
+						"User update failed. Modified date is in the future. ",
+						"Current date: ", new Date(), " Modified date: ",
+						modifiedDate));
```
[the remark this answers is not in the document]
Comment (markgulacsy) [reply modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/exportimport/LDAPUserImporterImpl.java:1753]: Thank you, I will extend the log message with the clockskew value

File: modules/apps/portal-security/portal-security-ldap-api/src/main/java/com/liferay/portal/security/ldap/exportimport/configuration/LDAPImportConfiguration.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (stian-sigvartsen) [reply modules/apps/portal-security/portal-security-ldap-api/src/main/java/com/liferay/portal/security/ldap/exportimport/configuration/LDAPImportConfiguration.java:39]: You are right, it is not a breaking change. The configured value is added to the current time and only if the LDAP `modifiedTimeStamp` is later (which is less likely than if the default `clockSkew` value is 0) will the exception be thrown and prevent login.

One thing I just realized is this new clockSkew is shared between all connected LDAP servers. Any clock difference is going to be unique to each LDAP server so can we move it to `LDAPServerConfiguration` ?

Also, I think we should add the `clockSkew` to `modifiedTimeStamp` prior to capping and assigning it to the user as `modifiedDate`?

**Pull request closed without merging**